File: Algorithm/greedy algorithm/atcoder/4-quadrant greedy method/B42/B42-gpt.go
// 以下は、**Go 1.20.6** に対応した、正確かつ高速な実装です。
// すべてのテストケースで正解となる「**4象限貪欲法**」を使い、関数分割・型明示・高速なI/Oを意識しています。

// ---

// ## ✅ 解法概要（再掲）

// * 各カード `(Ai, Bi)` に対して、4通りの方向（±Ai, ±Bi）でのスコア評価を行う。
// * 各方向で「貢献するカードのみ」を選び、スコア `|ΣAi| + |ΣBi|` を計算。
// * その中で最大のスコアを返す。

// ---

// ## ✅ Go実装（go 1.20.6）

// ```go
package main

import (
	"bufio"
	"fmt"
	"io"
	"math"
	"os"
	"strconv"
	"strings"
)

// Card 型: 各カードは表A・裏Bのペア
type Card struct {
	A int
	B int
}

/**
 * 最大スコアを計算（4象限貪欲法）
 * @param cards []Card - AとBを持つカードスライス
 * @return int - 最大スコア |ΣA| + |ΣB|
 */
func maxScore(cards []Card) int {
	maxScore := 0

	signs := [4][2]int{
		{1, 1},
		{1, -1},
		{-1, 1},
		{-1, -1},
	}

	for _, sign := range signs {
		sa, sb := sign[0], sign[1]
		sumA, sumB := 0, 0

		for _, c := range cards {
			va := sa * c.A
			vb := sb * c.B
			if va+vb > 0 {
				sumA += c.A
				sumB += c.B
			}
		}

		score := int(math.Abs(float64(sumA)) + math.Abs(float64(sumB)))
		if score > maxScore {
			maxScore = score
		}
	}

	return maxScore
}

/**
 * 標準入力から読み込み、カードをパースして最大スコアを出力
 * 処理時間: O(N), メモリ: O(N)
 */
func main() {
	cards := readInput(os.Stdin)
	result := maxScore(cards)
	fmt.Println(result)
}

/**
 * 入力読み込み関数
 * @param r io.Reader - 入力ストリーム
 * @return []Card - 読み込んだカードの配列
 */
func readInput(r io.Reader) []Card {
	scanner := bufio.NewScanner(r)
	scanner.Scan()
	n, _ := strconv.Atoi(scanner.Text())

	cards := make([]Card, 0, n)
	for i := 0; i < n && scanner.Scan(); i++ {
		parts := strings.Fields(scanner.Text())
		a, _ := strconv.Atoi(parts[0])
		b, _ := strconv.Atoi(parts[1])
		cards = append(cards, Card{A: a, B: b})
	}

	return cards
}

// ## ✅ 実行例

// 標準入力（input.txt）に以下を用意：

// ```
// 5
// 2 8
// 4 -5
// 5 -3
// -4 1
// -2 -3
// ```

// コマンドラインで：

// ```bash
// go run solve.go < input.txt
// ```

// 出力：

// ```
// 18
// ```

// ---

// ## ✅ 計算量と効率性

// | 項目     | 内容                            |
// | ------ | ----------------------------- |
// | 時間計算量  | **O(N)**（4方向スキャン）             |
// | メモリ計算量 | **O(N)**（カード格納）               |
// | I/O    | `bufio.Scanner` 使用による高速入力処理   |
// | 精度     | 正解率100%、浮動小数点誤差なし（`math.Abs`） |

// ---

// ## ✅ 型と構造の明示

// * `Card` 構造体：明確に `A` と `B` を分離
// * `maxScore`：すべてのロジックを関数にカプセル化
// * `readInput`：I/O 分離（テスト容易）

// ---

// ## ✅ 補足

// * 入力が `1e5` 件程度でもメモリ使用量は ≈ 1.6MB 以下（int×2 × 10万件）
// * `strconv.Atoi` は高速で安全
// * `math.Abs` は `float64` を取るため、`int→float64` キャストが必要（誤差は発生しません）

// ---

// 必要であれば：

// * `選んだカードのインデックスを出力する機能`
// * `高速読み込みの最適化 (fscanf)`
//   などの追加も可能です。

File: Algorithm/greedy algorithm/atcoder/4-quadrant greedy method/B42/B42-gpt.php
<!-- 以下は、**PHP 8.2.8** で動作する、正確かつ高速な解法です。
全ケースで正解となる **4象限貪欲法** を使い、関数化 + 型指定付き（PHPDocコメント）で記述しています。

---

## ✅ 実装方針（再確認）

スコアは `|ΣAi| + |ΣBi|`。
カードの選び方によって、最終的な `(ΣA, ΣB)` の長さを最大化すればよいため、

* **4象限の符号パターン** `(±1, ±1)` を使って、
* 各方向に「貢献するカードだけ」を選んで加算、
* 最終的なスコアを求めて、その中の最大を出力します。

---

## ✅ PHP 実装（PHP 8.2.8対応）

```php -->
<?php

/**
 * 与えられたカードの集合に対し、スコア |ΣA| + |ΣB| の最大値を求める。
 *
 * @param array<array{0: int, 1: int}> $cards 各カードの [A, B] ペア配列
 * @return int 最大スコア
 */
function maxScore(array $cards): int {
    $maxScore = 0;

    // 4つの象限に対応する符号の組み合わせ
    $signs = [
        [1, 1],
        [1, -1],
        [-1, 1],
        [-1, -1],
    ];

    foreach ($signs as [$sa, $sb]) {
        $sumA = 0;
        $sumB = 0;

        foreach ($cards as [$a, $b]) {
            $va = $sa * $a;
            $vb = $sb * $b;

            if ($va + $vb > 0) {
                $sumA += $a;
                $sumB += $b;
            }
        }

        $score = abs($sumA) + abs($sumB);
        if ($score > $maxScore) {
            $maxScore = $score;
        }
    }

    return $maxScore;
}

/**
 * 標準入力から読み込み、最大スコアを出力する
 * 処理時間: O(N)
 * メモリ使用量: O(N)
 */
function main(): void {
    $lines = explode("\n", trim(file_get_contents("php://stdin")));
    $n = (int)$lines[0];

    $cards = [];

    for ($i = 1; $i <= $n; $i++) {
        [$a, $b] = array_map('intval', explode(' ', trim($lines[$i])));
        $cards[] = [$a, $b];
    }

    $result = maxScore($cards);
    echo $result . PHP_EOL;
}

main();
// ```

// ---

// ## ✅ 実行例

// 標準入力に以下を与えると：

// ```
// 5
// 2 8
// 4 -5
// 5 -3
// -4 1
// -2 -3
// ```

// 実行結果：

// ```bash
// php solve.php < input.txt
// ```

// 出力：

// ```
// 18
// ```

// ---

// ## ✅ 計算量とメモリ効率

// | 項目        | 値                 |
// | --------- | ----------------- |
// | 時間計算量     | **O(N)**（4回走査）    |
// | 空間計算量     | **O(N)**（カード配列のみ） |
// | 型チェック・安全性 | PHPDocで明示         |

// ---

// ## ✅ 補足（PHP特有の注意点）

// * PHP は静的型を強制しないが、`array<array{int, int}>` のように **PHPDoc で型明示**すると IDE 補完や静的解析での安心感が得られます。
// * 演算コストやメモリ使用は `1e5` 件程度なら十分余裕があります（約2〜3MB）。

// ---

// ご希望があれば、`カードのインデックス出力` や `高速I/O処理（fscanf対応）` なども実装可能です。